Add preferences abstraction

faern opened the pull request:
PR #8 was growing quite a lot. So I felt like splitting out some of the code for separate review in order to shrink that PR a bit.

This PR introduces the safe abstraction for `SCPreferences`. The first commit is the code from @luozijun that I cherry picked from the other PR branch.  The second commit is where I add some simplifying constructors making it easier to construct the type for the most common use case.

@luozijun please add your feedback to this as well.

Please note that all constructors take `&CFString` instead of `&str`. This is in line with how we have done in `dynamic_store.rs` and also with how other crates implementing Apple APIs usually do. Example: https://docs.rs/core-graphics/0.13.0/core_graphics/font/struct.CGFont.html#method.from_name

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mullvad/system-configuration-rs/12)
<!-- Reviewable:end -->

LuoZijun commented:
@faern I think your implementation is pretty good.

When this PR got merged, i will rebase PR-8.

LuoZijun commented on system-configuration/src/preferences.rs:
```diff
@@ -13,8 +13,7 @@
 //! [`SCPreferences`]: https://developer.apple.com/documentation/systemconfiguration/scpreferences
```
Link: https://developer.apple.com/documentation/systemconfiguration/scpreferences-ft8

faern commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[system-configuration/src/preferences.rs, line 60 at r3](https://reviewable.io:443/reviews/mullvad/system-configuration-rs/12#-L7TIyfSbFEKQwevd38H:-L7TIyfSbFEKQwevd38I:bvvdlsx) ([raw file](https://github.com/mullvad/system-configuration-rs/blob/6ca71249ef8de7b682975d3cae83c5941ccf13e2/system-configuration/src/preferences.rs#L60)):*
> ```Rust
> 
>         unsafe {
>             SCPreferences::wrap_under_create_rule(SCPreferencesCreate(
> ```

I fixed a bug here now where the reference count was wrong. Since `wrap_under_get_rule` was used the retain count was incremented when it should not have been. Added test as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/system-configuration-rs/12)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[system-configuration/src/preferences.rs, line 13 at r2](https://reviewable.io:443/reviews/mullvad/system-configuration-rs/12#-L7RdFxXEDy3CG7-i7CL-r2-13:-L7TKrKuf70tJfCHQGRZ:bd5yf42) ([raw file](https://github.com/mullvad/system-configuration-rs/blob/e88ebcebf15b71c2ee105cc9082181a9e84db574/system-configuration/src/preferences.rs#L13)):*
<details><summary><i>Previously, LuoZijun (寧靜) wrote…</i></summary><blockquote>

Link: https://developer.apple.com/documentation/systemconfiguration/scpreferences-ft8
</blockquote></details>

Thanks. Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/system-configuration-rs/12)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 2 files at r1, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[system-configuration/src/preferences.rs, line 32 at r4](https://beta.reviewable.io/reviews/mullvad/system-configuration-rs/12#-L7VC9oT37MPTcJ3LUma:-L7VC9oT37MPTcJ3LUmb:bhf0jjp) ([raw file](https://github.com/mullvad/system-configuration-rs/blob/e78ac6f1f367dae7fdb34730876fdf4721daeb81/system-configuration/src/preferences.rs#L32)):*
> ```Rust
> 
> 
> impl SCPreferences {
> ```

Am I right that we do not implement any other methods except instantiation because we don't directly access them?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/system-configuration-rs/12)*
<!-- Sent from Reviewable.io -->

pronebird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/system-configuration-rs/12#-:-L7YynMg898FEiZoHNZZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/system-configuration-rs/12)*
<!-- Sent from Reviewable.io -->
